File: src/Bio.Frontend.Web/types/species.ts
/**
 * TypeScript types — Biodiversity Catalog (PostgreSQL)
 * Maps to: BioCommerce_Scientific database
 * Aligned with backend DTOs: SpeciesResponseDTO, SpeciesDetailDTO,
 * SpeciesListItemDTO, GeographicDistributionDTO, TaxonomyResponseDTO.
 */

/** TaxonomyResponse — mirrors TaxonomyResponseDTO */
export interface TaxonomyResponse {
    id: number;
    kingdom: string | null;
    phylum: string | null;
    className: string | null;
    orderName: string | null;
    family: string | null;
    genus: string | null;
}

/** SpeciesResponse — mirrors SpeciesResponseDTO (flat, no distributions) */
export interface SpeciesResponse {
    id: string;
    taxonomyId: number | null;
    taxonomy: TaxonomyResponse | null;
    slug: string;
    thumbnailUrl: string | null;
    scientificName: string;
    commonName: string | null;
    description: string | null;
    ecologicalInfo: string | null;
    traditionalUses: string | null;
    economicPotential: string | null;
    conservationStatus: string | null;
    altitudeRange: string | null;
    legalStatus: boolean;
    isSensitive: boolean;
    createdAt: string;
    updatedAt: string;
}

/** GeographicDistributionDTO — mirrors backend GeographicDistributionDTO */
export interface GeographicDistribution {
    id: string;
    speciesId: string;
    /** Latitude — null if species.isSensitive and user lacks privileged role */
    latitude: number | null;
    /** Longitude — null if species.isSensitive and user lacks privileged role */
    longitude: number | null;
    altitude: number | null;
    municipality: string | null;
    ecosystemType: string | null;
    /** True if coordinates were masked for bio-safety protection */
    isMasked: boolean;
}

/** RelatedProductDTO — lightweight product linked to a species via BaseSpeciesId */
export interface RelatedProduct {
    id: string;
    name: string;
    slug: string | null;
    description: string;
    price: number;
    stockQuantity: number;
    thumbnailUrl: string | null;
    isActive: boolean;
}

/** SpeciesDetailDTO — full detail response from GET /api/species/{id} or /slug/{slug} */
export interface SpeciesDetail {
    id: string;
    taxonomyId: number | null;
    taxonomy: TaxonomyResponse | null;
    slug: string;
    thumbnailUrl: string | null;
    scientificName: string;
    commonName: string | null;
    description: string | null;
    ecologicalInfo: string | null;
    traditionalUses: string | null;
    economicPotential: string | null;
    conservationStatus: string | null;
    altitudeRange: string | null;
    legalStatus: boolean;
    isSensitive: boolean;
    createdAt: string;
    updatedAt: string | null;
    distributions: GeographicDistribution[];
    relatedProducts: RelatedProduct[];
}

/** SpeciesListItemDTO — lightweight DTO for catalog grid/list */
export interface SpeciesListItem {
    id: string;
    slug: string;
    scientificName: string;
    commonName: string | null;
    thumbnailUrl: string | null;
    conservationStatus: string | null;
    isSensitive: boolean;
    kingdom: string | null;
    family: string | null;
    createdAt: string;
}

/** Species search/filter params — mirrors backend SpeciesFilterParams */
export interface SpeciesSearchParams {
    query?: string;
    kingdom?: string;
    phylum?: string;
    className?: string;
    orderName?: string;
    family?: string;
    genus?: string;
    isSensitive?: boolean;
    conservationStatus?: string;
    page?: number;
    pageSize?: number;
    sortBy?: "scientificName" | "commonName" | "createdAt";
    sortOrder?: "asc" | "desc";
}

/** SpeciesFilterMetaDTO — distinct filter values from backend */
export interface SpeciesFilterMeta {
    kingdoms: string[];
    phylums: string[];
    classes: string[];
    orders: string[];
    families: string[];
    genera: string[];
    conservationStatuses: string[];
    totalSpeciesCount: number;
}

File: src/Bio.Backend.Core/Bio.Domain/Entities/Species.cs
namespace Bio.Domain.Entities;

/// <summary>
/// Especie del catálogo de biodiversidad.
/// Tabla: species (PostgreSQL).
/// </summary>
public class Species
{
    public Guid Id { get; private set; }
    public int? TaxonomyId { get; private set; }
    public string Slug { get; private set; } = string.Empty;
    public string? ThumbnailUrl { get; private set; }
    public string ScientificName { get; private set; } = string.Empty;
    public string? CommonName { get; private set; }
    public string? Description { get; private set; }
    public string? EcologicalInfo { get; private set; }
    public string? TraditionalUses { get; private set; }
    public string? EconomicPotential { get; private set; }
    public string? ConservationStatus { get; private set; }
    public bool IsSensitive { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime UpdatedAt { get; private set; }

    public Taxonomy? Taxonomy { get; private set; }
    public ICollection<GeographicDistribution> GeographicDistributions { get; private set; } = new List<GeographicDistribution>();

    private Species() { }

    public Species(
        Guid id,
        string slug,
        string scientificName,
        int? taxonomyId = null,
        string? thumbnailUrl = null,
        string? commonName = null,
        string? description = null,
        string? ecologicalInfo = null,
        string? traditionalUses = null,
        string? economicPotential = null,
        string? conservationStatus = null,
        bool isSensitive = false)
    {
        Id = id;
        TaxonomyId = taxonomyId;
        Slug = slug;
        ThumbnailUrl = thumbnailUrl;
        ScientificName = scientificName;
        CommonName = commonName;
        Description = description;
        EcologicalInfo = ecologicalInfo;
        TraditionalUses = traditionalUses;
        EconomicPotential = economicPotential;
        ConservationStatus = conservationStatus;
        IsSensitive = isSensitive;
        CreatedAt = DateTime.UtcNow;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Update(
        string? slug,
        string? thumbnailUrl,
        string? commonName,
        string? description,
        string? ecologicalInfo,
        string? traditionalUses,
        string? economicPotential,
        string? conservationStatus,
        bool? isSensitive,
        int? taxonomyId)
    {
        if (slug != null) Slug = slug;
        if (thumbnailUrl != null) ThumbnailUrl = thumbnailUrl;
        if (commonName != null) CommonName = commonName;
        if (description != null) Description = description;
        if (ecologicalInfo != null) EcologicalInfo = ecologicalInfo;
        if (traditionalUses != null) TraditionalUses = traditionalUses;
        if (economicPotential != null) EconomicPotential = economicPotential;
        if (conservationStatus != null) ConservationStatus = conservationStatus;
        if (isSensitive.HasValue) IsSensitive = isSensitive.Value;
        if (taxonomyId.HasValue) TaxonomyId = taxonomyId;
        UpdatedAt = DateTime.UtcNow;
    }
}
